cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternalsCommon.qll

```diff
@@ -320,10 +320,12 @@ private module IteratorIndirections {
   }
 }
 
-predicate isDereference(Instruction deref, Operand address) {
-  any(Indirection ind).isAdditionalDereference(deref, address)
+predicate isDereference(Instruction deref, Operand address, boolean additional) {
```

Comment:
I think it's about time this growing predicate got a brief QLDoc comment.

What is the intuition behind caring whether a deref is an actual `LoadInstruction` or 'additional'?

Reply:
I've added QLDoc to this predicate in 261ba8e02d9ebe13fb9a0bc0d495d916df84efd4.

The reason we sometimes need to care about whether a deref is a `LoadInstruction` or 'additional' is that, when the dereference is a `LoadInstruction` then the dataflow node that represents the indirection of the address of the load represents _exactly_ the same semantic value as the result of the load. i.e., in the example from the PR description:
```
r1(glval<char **>) = VariableAddress[a]          :
r2(char **)        = Load[a]                     : &:r1, m1
```
if we have an indirect dataflow node for `VariableAddress[a]`, then this represents the same value as the result of the load. And in those cases we can "share" their representation by the same `DataFlow::Node`.

In contrast, when the load is an additional (such as an iterator's or smart pointer's `operator*`) then they don't really represent the same value since the indirection of the address operand has an iterator (or smart pointer) type, whereas the result of calling `operator*` will give a value of whatever type the iterator (or smart pointer) wraps.

For example in something like:
```
std::shared_ptr<int> p = ...;
int x = *p; // this is an additional load. The indirection of the address dataflow node of type `std::shared_ptr<int>`, but the result of the load has type `int`.
```

Reply:
That makes sense, and thanks for adding the QLDoc.